fb-negotiation/neural/model_builder.py

```diff
@@ -0,0 +1,342 @@
+    num_word_embeddings = len(word_dict)
+
+    return Embeddings(word_vec_size=embedding_dim,
+                      position_encoding=False,
+                      dropout=opt.dropout,
+                      word_padding_idx=word_padding_idx,
+                      word_vocab_size=num_word_embeddings)
+
+
+def make_encoder(opt, embeddings):
+    """
+    Various encoder dispatcher function.
+    Args:
+        opt: the option in current environment.
+        embeddings (Embeddings): vocab embeddings for this encoder.
+    """
+    if opt.encoder_type == "transformer":
+        return TransformerEncoder(opt.enc_layers, opt.rnn_size,
+                                  opt.dropout, embeddings)
+    elif opt.encoder_type == "cnn":
+        return CNNEncoder(opt.enc_layers, opt.rnn_size,
+                          opt.cnn_kernel_width,
+                          opt.dropout, embeddings)
+    elif opt.encoder_type == "mean":
+        return MeanEncoder(opt.enc_layers, embeddings)
+    else:
+        # "rnn" or "brnn"
+        bidirectional = True if opt.encoder_type == 'brnn' else False
+        return StdRNNEncoder(opt.rnn_type, bidirectional, opt.enc_layers,
+                        opt.rnn_size, opt.dropout, embeddings)
+
+def make_selectors(opt, kb_dict):
+    selectors = {}
+    '''
+    Create selection encoder:
+        - For FB, this combines attention output and context hidden
+        - For Cocoa, this combines decoder output and kb scenario output
+    '''
+    input_size = opt.rnn_size + opt.kb_embed_size
+    selectors["enc"] = nn.Sequential(
+        torch.nn.Linear(input_size, opt.sel_hid_size),
+        nn.Tanh()
+    )
+    '''
+    Create selection decoders, one per each item. Range is 6 (and not 3)
+        because we are training the model to keep track of not only its
+        own selection but also the partner's predicted selection.
+    Items are [book, hat, ball] in exactly that order, so that we match
+        the order found in the training data. Selectors 1 to 3 are for
+        "my selection", selectors 4 to 6 are for "partner's selection".
+    '''
+    selectors["dec"] = nn.ModuleList()
+    for i in xrange(6):
+        selectors["dec"].append(nn.Linear(opt.sel_hid_size, kb_dict.size))
+
+    return selectors
+
+def make_context_embedder(opt, embeddings, embed_type='utterance'):
+    """
+    Various context embedder dispatcher function. See make_encoder for options
+    embed_type (:str:): either dialogue, kb (for title and desc) or category
+    """
+    if opt.context_embedder_type == "mean":
+        return MeanEncoder(opt.enc_layers, embeddings, embed_type)
+    else:
+        # "rnn" or "brnn"
+        bidirectional = True if opt.context_embedder_type == 'brnn' else False
+        return StdRNNEncoder(opt.rnn_type, bidirectional, opt.enc_layers,
+                        opt.rnn_size, opt.dropout, embeddings, embed_type, False)
+
+def make_decoder(opt, embeddings, tgt_dict):
+    """
```

Comment:
import form cocoa

Reply:
this one you need some override. but you can always call the cocoa one first, and add whatever you need after. it.